### DISCUSSION TITLE
input plugin: custom commands

### DISCUSSION THREAD

**Discussion by hhirtz** [MusicPlayerDaemon/MPD#1330] [Ideas]
Idea: allow users to define custom commands as input plugins, e.g. with this configuration

```
input_command {
        scheme "http"
        scheme "https"
        command "yt-dlp"
}
```

and this input:

`mpc add https://www.youtube.com/watch?v=DON5WvSVYSc`

mpd would spawn [`yt-dlp`][0] with the URL as first argument, or something, and read the audio from its standard output (the curl input plugin would have a lower priority for this example to work.. or be disabled completely in the configuration).

This could be an easier way to support the following ideas, without adding more lines of code to this project:
- youtube plugin https://github.com/MusicPlayerDaemon/MPD/pull/223 https://github.com/MusicPlayerDaemon/MPD/pull/955
- qobuz plugin https://github.com/MusicPlayerDaemon/MPD/issues/1209
- tidal plugin https://github.com/MusicPlayerDaemon/MPD/issues/627
- youtube-dl and yt-dlp have a retry mechanism already https://github.com/MusicPlayerDaemon/MPD/issues/431
- ... and other things, maybe.. haven't looked much at the issue tracker

I'm willing to implement such plugin, and wanted to know your thoughts about this before digging into the code base. 😃 

[0]: https://github.com/yt-dlp/yt-dlp

**Comment by MaxKellermann** [comment]
If you believe the feature is useful (I don't need it), and if you do it properly, I'll merge it. Beware, there are lots of complexities in what looks simple at first:
- will you support seeking?
- will you support indicating the total size?
- will you support indicating the MIME type?
- will you support tags?
- how do you propagate error conditions between the child process and MPD?
- will you support Windows?
- spawning child processe from MPD is difficult; there are things like the signal block mask that need to be considered. I can help with that part.

**Reply by hhirtz** [reply]
Thanks for the quick reply.  I'll look into these points and come back when I have something to show.
